src/client/api/rpc.c

```diff
@@ -114,16 +113,10 @@ query_cb(struct crt_proto_query_cb_info *cb_info)
 	int			 rc;
 
 	if (daos_rpc_retryable_rc(cb_info->pq_rc)) {
-		rc = rsvc_client_choose(&rproto->cli, &rproto->ep);
-		if (rc) {
-			D_ERROR("rsvc_client_choose() failed: "DF_RC"\n", DP_RC(rc));
-			rproto->rc = rc;
-			rproto->completed = true;
-		}
-
-		rc = crt_proto_query_with_ctx(&rproto->ep, rproto->base_opc,
-					      rproto->ver_array, rproto->array_size,
-					      query_cb, rproto, daos_get_crt_ctx());
+		rproto->ep.ep_rank = (rproto->ep.ep_rank + 1) % rproto->nr_ranks;
```

Comment:
@frostedcmos does this function return the list of alive ranks? how does it track the alive ranks?

Reply:
hmm.. thinking on it more, perhaps on a client that's not the right call to use.
This function returns the list of ranks that it knows uris for, however clients are allowed to send RPCs to ranks with no uris (in which case uri lookup is done by cart automatically).

cart provides apis to add/remove/modify group list (crt_group_primary_modify()/crt_group_secondary_modify()), but the caller is the one who needs to invoke those. in engine case removal is done by swim for example. 

in clients case, looks like group is only populated during the attach response, but not managed afterwards:
 789         for (i = 0; i < resp->n_rank_uris; i++) {
 790                 Mgmt__GetAttachInfoResp__RankUri *rank_uri = resp->rank_uris[i];
 791
 792                 rc = crt_group_primary_rank_add(daos_get_crt_ctx(), group,
 793                                                 rank_uri->rank, rank_uri->uri);

Reply:
OK i think in that case, there i nothing else we can do for this other than a random ping on all possible ranks, because the client at this init point has no information on the system that it maintained anywhere; there is no pool map here or anything. and as Kris has mentioned in another channel, the agent does not keep track of ranks that are excluded, so if by chance we ping a dead rank, we just have to retry.